lnrpc/routerrpc/router_server.go

```diff
@@ -398,14 +441,350 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	// have to add it to the original payment amount. Only then will the
+	// payment probe be able to determine the correct fee to the last hop
+	// prior to the private destination. For example, if the user wants to
+	// send 1000 sats to a private destination and the last hop's fee is 10
+	// sats, then 1010 sats will have to arrive at the last hop. This means
+	// that the probe has to be dispatched with 1010 sats to correctly
+	// calculate the routing fee.
+	//
+	// Calculate the hop fee for the last hop manually.
+	hopFee := lspHopHint.HopFee(*payReq.MilliSat)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add the last hop's fee to the requested payment amount that we want
+	// to get an estimate for.
+	probeRequest.AmtMsat += int64(hopFee)
+
+	// Dispatch the payment probe with adjusted fee amount.
+	resp, err := s.sendProbePayment(ctx, probeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// The probe succeeded, so we can add the last hop's fee to fee the
+	// payment probe returned.
+	resp.RoutingFeeMsat += int64(hopFee)
+
+	lastHopTimeLock := int64(lspHopHint.CLTVExpiryDelta)
+	if lastHopTimeLock < int64(payReq.MinFinalCLTVExpiry()) {
+		lastHopTimeLock = int64(payReq.MinFinalCLTVExpiry())
+	}
+	resp.TimeLockDelay += lastHopTimeLock
+
+	return resp, nil
+}
+
+// isLSP checks if the route hints indicate an LSP. An LSP is indicated in the
+// first return value if the last node in each route hint has the same node id.
+// If an LSP is indicated, the remaining hops in each route hint are returned
+// to the caller to be considered by the caller as route hints to the LSP node.
+// Additionally, the identified LSP node is returned as a separate hop hint.
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to heuristically determine a worst-case
+// estimate.
+func isLSP(routeHints [][]zpay32.HopHint, amt lnwire.MilliSatoshi) (bool,
+	[][]zpay32.HopHint, *zpay32.HopHint) {
+
+	if len(routeHints) == 0 || len(routeHints[0]) == 0 {
+		return false, [][]zpay32.HopHint{}, nil
+	}
+
+	// In case we are dealing with an LSP, we construct a modified list of
+	// route hints that allows the caller to probe the LSP, which itself is
+	// returned as a separate hop hint.
+	adjustedRouteHints := make([][]zpay32.HopHint, 0, len(routeHints))
+
+	// If the final pubkey of each route hint is the same, we can assume
+	// that this identifies an LSP node. The first candidate is our
+	// reference hop hint.
+	refHopHint := routeHints[0][len(routeHints[0])-1]
+
+	// Strip off the potential LSP hop hint from the reference route hint,
+	// and add the remaining hop hints to the adjusted route hints.
+	if len(routeHints[0]) > 1 {
+		adjustedRouteHints = append(
+			adjustedRouteHints,
+			routeHints[0][:len(routeHints[0])-1],
+		)
+	}
+	for i := 1; i < len(routeHints); i++ {
+		rh := routeHints[i]
+
+		// Skip empty route hints.
+		if len(rh) == 0 {
+			continue
+		}
```

Comment:
I don't see where this would have an impact. It indicates that the invoice creator messed something up for a single hint, as long as the other hints are functional we can go ahead and estimated imo.